Show var source error on resource and build page

xtremerui opened the pull request:
## Changes proposed by this PR

On build page, add 'do' step to step trees so error event could
be linked to it correctly.

<img width="626" alt="Screen Shot 2022-01-27 at 11 59 55 AM" src="https://user-images.githubusercontent.com/1585949/151406941-f09c489d-6f51-4951-ad7b-13295059196f.png">

On resource page, the check was created async and hoping backend
to finish the check build. But if var source failed in first, the
check build was never be able to run the `check` script. Now after
the check request, ui will ask for check build plan explicitly to
get the errored build events.

<img width="647" alt="Screen Shot 2022-01-27 at 11 59 44 AM" src="https://user-images.githubusercontent.com/1585949/151406916-e40eeb09-d8d5-4c8a-9964-ee2e5d2fbf99.png">

Note: the auto ticking to udpate resource check status relys on if the resource having a build linked. In this case, the check build errored on the first stage by invalid var interpolation, so the build is never ran and linked to the resource. One will need to manually trigger the check to see to error.


closes #7927

* [x] done
* [ ] todo

## Release Note
 * Now error caused by variable interpolation can be shown correctly on resource and build page.

Commit:
show var source error on resource and build page

On build page, add 'do' step to step trees so error event could
be linked to it correctly.

On resource page, the check was created async and hoping backend
to finish the check build. But if var source failed in first, the
check build was never be able to run the `check` script. Now after
the check request, ui will ask for check build plan explicitly to
get the errored build events.

Signed-off-by: Rui Yang <ruiya@vmware.com>

clarafu commented:
Sorry, I just wanted to confirm with you that I'm understanding this properly. Is the fix here that the user will have to manually check the resource in order to see the error? Is there no way we can show the error without having the user need to manually run a check?

I think it would be best if we can expose the error to the user without requiring them to manually check the resource since this is how it worked before the 7.6.0 changes that caused this bug. I think all the parts are there but we just need to modify some stuff to make it save the in memory build events in the database and then tie it to the resource using the `in_memory_build_id` on the resource/resourcetype/prototypes column. 

There must be something different between a regular build and an in memory build where we don't tie the build to the resource until after the resource check is run/finished. Because if I go to a version of concourse prior to 7.6 (without the in memory check changes), I can see that the error appears because the check build is tied to the resource through the `build_id` column on the resources table and we save the error into the `check_build_events` table when the var fails to interpolate or whenever a build fails to run during the setup stage. But for in memory checks (post 7.6) the `in_memory_build_id` column is empty on the resources table when a check fails so there is no way to tie the check build events that saves the error to the resource check build.

xtremerui commented:
Yeah I noticed the comment in 
https://github.com/concourse/concourse/blob/4ec23fd7c6011d3ba6be7f6ba019d411ec53129e/atc/exec/check_step.go#L180-L183

xtremerui commented:
@clarafu I have done a deep dive into this issue and found the root cause. I am writing up a summary.

Commit:
undo UI change in resource page

since it is a 100% backend issue.

In this commit it init check build events earlier than the source
interpolation so when var source fails the error events could be
recorded.

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented:
I have pushed a commit that *partly* fixes the resource page var source error displaying issue. But there are caveats.

 1 . By now with the latest fix, the error will show properly as normal. However, the error will be flicking every 5-15 secs due to the check running in server. The reason is when a check in memory happens and it fails var source interpolation, there is no resource config scope created the the check. So there wont be any build infos like it used to have in the resource query

https://github.com/concourse/concourse/blob/223485fe795e1eb9e2bc3892ce3bcc878c1cc4a2/atc/db/resource.go#L130-L134

everytime scanner ticks, it will consider the check never ran and it will create the check again. From the UI you then see  the resource state is flicking and refreshing every 5-15 sec or so as frontend is quering backend for resource and check build info.

2. In the pipeline view, a resource with var source error won't be shown as yellow. Same reason as 1, since there is no resource config scope created for such resource check, the check build status is lost. In fact, for all in memory check builds, there is no buidl status afterwards. It is inferred from a logic here

https://github.com/concourse/concourse/blob/223485fe795e1eb9e2bc3892ce3bcc878c1cc4a2/atc/db/resource.go#L977-L988

thus for a check build that failed by interpolation (that only build start time is available), the check build status read from UI will be always `started` (versus the error status shown in resource page comes from the build events status).

For caveat 1, I don't think it is really a bug since thats how it works.
For caveat 2, to fix it we have to add a new column `in_memory_build_end_time` to `resource` table. Then we could apply similar method as shown in 2 to determine a check build status to be errored if it only has start time but no end time (as var source interpolation failed in between).

I am not sure if caveat 2 worth a fix that needs migration in this case.

clarafu commented:
Great job having the error message be able to be shown even without manually triggering! I just tested it out and it works great, the refreshing/flickering of the error message is not a problem at all. For the second problem, I'm not sure if just an end time will work because if the check is currently running then it will have a started time and not end time. I feel like we would need something more like a column for status on the resources table. We could think about adding that rather than doing all this extra logic here https://github.com/concourse/concourse/blob/223485fe795e1eb9e2bc3892ce3bcc878c1cc4a2/atc/db/resource.go#L977-L1013
I think most of that logic is for coming up with a build status because I think we were trying to avoid needing to update the resources table after an in memory check has finished. But I can't think of any way around it if we want to show the proper status 🤔

Commit:
Add build status column in resources table

so it can reflect the in-memory build status correctly.
when in-memory build is created, its status is pending.
once engine picks the build to run, its status becomes
started.

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented on atc/db/build_in_memory_check.go:
```diff
@@ -206,6 +206,7 @@ func (b *inMemoryCheckBuildForApi) Events(from uint) (EventSource, error) {
 					err = psql.Select("1").
 						From("resources").
 						Where(sq.Eq{"id": b.resourceId}).
+						Where(sq.Eq{"in_memory_build_status": BuildStatusStarted}).
```
Without this line, the build events of  a resource that is with var source error will not be closed from server, so UI will keep showing `loading ...` . The reason is no resource config scope was created due to var error and this query will actually find one record, thus it will return completed as false.
https://github.com/concourse/concourse/blob/4e6ad40006517d4dcf61cc8146125bf82831456f/atc/db/build_in_memory_check.go#L215

evanchaoli commented on atc/exec/check_step.go:
```diff
@@ -99,6 +99,10 @@ func (step *CheckStep) run(ctx context.Context, state RunState, delegate CheckDe
 		"step-name": step.plan.Name,
 	})
 
+	if err := delegate.BeforeSelectWorker(logger); err != nil {
```
I have concern to pull up `BeforeSelectWorker` to before `delegate.Initializing()`. Looking at

https://github.com/concourse/concourse/blob/fa6ef67835be414278cba484305a8e656ce7a9b3/atc/engine/build_step_delegate.go#L171-L185

`BeforeSelectWorker` means a worker is needed, which implied a real check needs to run. So when only when an in-memory-check really needs to run check, it interacts with DB. If pull up `BeforeSelectWorker`, then all in-memory-check will init db, which will completely eliminate performance improvement of in-memory-check.

Commit:
init build events when build finishing

in this way we dont need to interact with DB until there is an
actual build event that needs to be recorded

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented on atc/db/build_in_memory_check.go:
```diff
@@ -416,10 +419,6 @@ func (b *inMemoryCheckBuild) AcquireTrackingLock(logger lager.Logger, interval t
 }
 
 func (b *inMemoryCheckBuild) Finish(status BuildStatus) error {
-	if !b.runningInContainer {
```
@evanchaoli I think your concern is valid so I have updated it to only init the DB while it indeed need to record the build event.

The only thing here is `runningInContainer`, while an error like var source interpolation happens, that value here is false (since the build hasn't really being ran in a container yet). So I have to remove this check to allow it to init the DB.

evanchaoli replied:
If you remove `runningInContainer`, then all non-run checks will also write to db, which still doesn't resolve my concern. I think you can do something like:

```golang
if !b.runningInContaienr && status == SUCCEED
```

xtremerui replied:
In engine.go it only calls b.finish when there is error that needs to be recorded. Why `all non-run checks` would call finish since they haven't started to run yet? Is there somewhere else that method is used?

xtremerui replied:
Also, `Finish` clean up entries in containers and resource caches table, if we gate it by `succeeded`, what will happen to the containers and caches used for the succeeded check build?

evanchaoli replied:
We don't simply gate it by `succeeded`. When 'runningInContainer` is true, it's not gated, so clean up will still run.

xtremerui replied:
I see. So if `runningInContainer` is false and it is succeeded, it means the check build finds a version so it doesn't need to create a container to run the check, and thus it won't need to run clean up in `Finish`.

evanchaoli commented on atc/db/build_in_memory_check_test.go:
```diff
@@ -245,6 +254,14 @@ var _ = Describe("Build", func() {
+
+						Expect(defaultResource.BuildSummary().Status).To(Equal(atc.StatusSucceeded))
+					})
+
```
I think we need to add two test cases when `OnCheckBuildStart` is not called:

1. Finish with "succeeded", no db operation should be done, this is for non-run checks.
2. Finish with an error, db operation should be done, this is for the problem this pr is fixing.

xtremerui replied:
I am a bit confused, when you say `non-run checks` do you mean a succeeded check build that saves its build events to memory? If yes, call it `non-db checks` might be better.

`non-run checks` to me means a check build that picked up by engine but failed to run the actual `check` script in `exec/check_step.go`, like the check with var error in this case.

evanchaoli replied:
I meant the cases where `checkDelegate.WaitToRun()` returns false, thus no need to start a container and run check.

xtremerui replied:
> `checkDelegate.WaitToRun()` returns false

I suppose you meant this here
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/exec/check_step.go#L146

if `run` is fales, it goes to else block to store the found version and then
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/exec/check_step.go#L233

since check delegate doesn have an implementation of `Finished`, it goes to the one in build step delegate
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/engine/build_step_delegate.go#L151

which will then save the finish event to the build using `SaveEvent` here
https://github.com/concourse/concourse/blob/0aeac8807b40acfcc9b806ab37e6286cc16ed785/atc/db/build_in_memory_check.go#L518

I might be missing some important part, but the non-run checks you meant are not relevant to the `Finish` method that I changed. Could you give some more details?

Commit:
dont init DB if it is a succeeded in memory check build

Signed-off-by: Rui Yang <ruiya@vmware.com>

evanchaoli commented on atc/engine/check_delegate.go:
```diff
@@ -70,6 +70,13 @@ func (d *checkDelegate) Initializing(logger lager.Logger) {
 	logger.Info("initializing")
 }
 
+func (d *checkDelegate) BeforeSelectWorker(logger lager.Logger) error {
```
Why this is needed? It has been implemented in build_step_delegate.

xtremerui replied:
Ah it was added by previous commit that for init DB before var source interpolation. It is not needed anymore.

clarafu reviewed (changes requested):
I tested this out manually and it all works really nicely! I tested it out with the following cases:
- a resource with failing var in source config
- a resource sharing the same resource config (with global resources) + failing var
- a resource type with failing var

The only kind of weird thing that I don't think is a problem is that the error from the resource type var interpolation shows up in the resource's error

![image](https://user-images.githubusercontent.com/19290703/160193061-7142e04e-1233-43df-81bc-5d2ba0be3067.png)

Other than that, I think after we clean up the code and also adding those test cases that Evan mentioned, it should be good to merge!

xtremerui commented on atc/db/build_in_memory_check_test.go:
```diff
@@ -126,6 +127,46 @@ var _ = Describe("Build", func() {
 				Expect(err.Error()).To(Equal("no build event"))
 			})
 
+			Context("BeforeCheckBuildStart", func() {
```
@clarafu these are the two tests that evanchaoli mentioned.

Commit:
cleanup tests

Signed-off-by: Rui Yang <ruiya@vmware.com>

clarafu commented on atc/db/resource_test.go:
```diff
@@ -195,7 +195,6 @@ var _ = Describe("Resource", func() {
 					Expect(buildSummary.Name).To(Equal(db.CheckBuildName))
 					Expect(buildSummary.TeamName).To(Equal(scenario.Team.Name()))
 					Expect(buildSummary.PipelineName).To(Equal(scenario.Pipeline.Name()))
-					Expect(buildSummary.Status).To(Equal(atc.StatusFailed))
```
Should we be testing it to return the status that is in the db? (Pending build?)

xtremerui replied:
There is no such status in db since the build (id 99) never exists. The status in the removed line comes from 
https://github.com/concourse/concourse/blob/a188c2a9d671597cd3f724807800731a6df563be/atc/db/resource.go#L1003

and 

https://github.com/concourse/concourse/blob/a188c2a9d671597cd3f724807800731a6df563be/atc/db/resource.go#L133

since in the test when initializing the resource config scope the field never got set to true. And there is no proper way to set it othan than DB surgery..
